vault/ssorolecredentialsprovider.go

```diff
@@ -201,3 +211,191 @@ func (p *SSORoleCredentialsProvider) newOIDCToken(ctx context.Context) (*ssooidc
 		return t, nil
 	}
 }
+
+// newOIDCTokenPKCE generates a new OIDC token using the "Authorization Code Grant" flow with PKCE.
+func (p *SSORoleCredentialsProvider) newOIDCTokenPKCE(ctx context.Context) (*ssooidc.CreateTokenOutput, error) {
+	// ref: https://datatracker.ietf.org/doc/html/rfc7636
+
+	// generate a random 32 byte code verifier; base64 encode it
+	codeVerifierBytes := make([]byte, 32)
+	n, err := crand.Read(codeVerifierBytes)
+	if err != nil || n != 32 {
+		return nil, fmt.Errorf("failed to generate PKCE verifier: %w", err)
+	}
+	codeVerifier := base64.RawURLEncoding.EncodeToString(codeVerifierBytes)
+
+	// generate the code challenge: base64(sha256(codeVerifier))
+	codeChallengeBytes := sha256.Sum256([]byte(codeVerifier))
+	codeChallenge := base64.RawURLEncoding.EncodeToString(codeChallengeBytes[:])
+	log.Printf("Generated PKCE code_challenge: %q", codeChallenge)
+
+	clientCreds, err := p.OIDCClient.RegisterClient(ctx, &ssooidc.RegisterClientInput{
+		ClientName:   aws.String("aws-vault"),
+		ClientType:   aws.String("public"),
+		GrantTypes:   []string{"authorization_code", "refresh_token"},
+		Scopes:       []string{"sso:account:access"},
+		IssuerUrl:    aws.String(p.StartURL),
+		RedirectUris: []string{"http://127.0.0.1/oauth/callback"},
+	})
+	if err != nil {
+		return nil, err
+	}
+	log.Printf("Created new OIDC client (expires at: %s)", time.Unix(clientCreds.ClientSecretExpiresAt, 0))
+
+	// start the callback server
+	cbServer, err := newOauthCallbackServer()
+	if err != nil {
+		return nil, fmt.Errorf("failed to create oauthCallbackServer: %w", err)
+	}
+	log.Printf("oauthCallbackServer callback endpoint: %s", cbServer.redirectURI())
+	go func() {
+		if err := cbServer.Serve(); err != nil && !errors.Is(err, http.ErrServerClosed) {
+			log.Printf("Failed to run oauthCallbackServer: %s", err)
+		}
+	}()
+	// keep a copy of the redirectURI for the CreateToken call (as the server will be closed after the code is received)
+	redirectURI := cbServer.redirectURI()
+
+	// construct the authorize URL with the client and PKCE parameters
+	args := url.Values{
+		"client_id":             {aws.ToString(clientCreds.ClientId)},
+		"response_type":         {"code"},
+		"redirect_uri":          {redirectURI},
+		"state":                 {cbServer.state},
+		"code_challenge_method": {"S256"},
+		"code_challenge":        {codeChallenge},
+		"scopes":                {"sso:account:access"},
+	}
+	// prefer the base endpoint from client options, otherwise use a default
+	var host string
+	if p.OIDCClient.Options().BaseEndpoint != nil && *p.OIDCClient.Options().BaseEndpoint != "" {
+		host = *p.OIDCClient.Options().BaseEndpoint
+	} else {
+		host = "oidc.us-east-1.amazonaws.com"
+	}
```

Comment:
Hardcoding the region to `us-east-1` won't work in all use cases. For example our SSO region is `eu-west-1`.

To get this to work I used `host = fmt.Sprintf("oidc.%s.amazonaws.com", p.OIDCClient.Options().Region)`